src/Analyser/MutatingScope.php

```diff
@@ -1621,12 +1621,18 @@ private function resolveType(string $exprString, Expr $node): Type
 			if ($this->hasExpressionType($node)->yes()) {
 				return $this->expressionTypes[$exprString]->getType();
 			}
-			return $this->initializerExprTypeResolver->getClassConstFetchTypeByReflection(
+			$constantType = $this->initializerExprTypeResolver->getClassConstFetchTypeByReflection(
 				$node->class,
 				$node->name->name,
 				$this->isInClass() ? $this->getClassReflection() : null,
 				fn (Expr $expr): Type => $this->getType($expr),
 			);
+
+			if ($this->isInTrait() && !$constantType->isLiteralString()->yes()) {
```

Comment:
literals are not generalized, to make sure class-strings still work like expected.

see existing test at

https://github.com/phpstan/phpstan-src/blob/f4cfe25b25e6ee26d7227f20dc754e11d6de2ee4/tests/PHPStan/Analyser/data/type-aliases.php#L170

Reply:
This is wrong assumption and a wrong fix. Sure, `Test::FOO` (including `::class`) should not change, but that has nothing to do with literal strings.

Instead, you need to look at the class the constant is fetched on. If it's the current class and we're in a trait, we essentially need to make it `mixed`... But if it's fetched from an external place, we can keep it as it is.

Of course, `self::class` doesn't need to be mixed, but should be kept as `class-string&literal-string`.